humility-net-core/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+        {
+            if base <= addr && base + size > addr {
+                let start = (addr - base) as usize;
+                let roffs = offset + start;
+
+                if start + data.len() <= size as usize {
+                    data.copy_from_slice(
+                        &self.flash.contents[roffs..roffs + data.len()],
+                    );
+
+                    return Ok(());
+                }
+
+                let len = (size as usize) - start;
+                data[..len]
+                    .copy_from_slice(&self.flash.contents[roffs..roffs + len]);
+
+                return self.read(addr + len as u32, &mut data[len..]);
+            }
+        }
+
+        self.read_ram(addr, data).context(format!(
+            "0x{addr:0x} can't be read via the archive or over the network"
+        ))
+    }
+
+    /// Reads data from RAM using the (UDP) dump agent as a proxy
+    fn read_ram(&mut self, addr: u32, data: &mut [u8]) -> Result<()> {
+        let p = self.ram.partition_point(|v| v.base <= addr);
+        if p == 0 {
+            bail!("address not found in RAM map");
+        }
+
+        // By construction, we know that every region in self.ram has at least
+        // one task associated with it.  If this task is the supervisor, then
+        // that's bad news.
+        let region = &self.ram[p - 1];
+        let Some(&task) = region.tasks.iter().find(|t| t.task() != 0) else {
+            bail!("supervisor memory cannot be read using dump facilities");
+        };
+
+        let mut agent_core = DumpAgentCore::new(self.flash.clone());
+        let mut udp_dump = UdpDumpAgent::new(self);
+
+        // Send a request to dump this specific region of memory within the task
+        let aligned_start = addr & !0b11;
+        let mut aligned_length = data.len() as u32 + addr - aligned_start;
+        while aligned_length & 0b11 != 0 {
+            aligned_length += 1;
+        }
+
+        let dump_index: u8 = match udp_dump.dump_task_region(
+            task.task(),
+            aligned_start,
+            aligned_length,
+        ) {
+            Ok(addr) => addr,
+            Err(e) if e.to_string().contains("DumpAreaInUse") => {
+                bail!(
+                    "out of space for dump_agent; use \
+                     `humility dump --initialize-dump-agent` to free space"
+                )
+            }
+            Err(e) => {
+                bail!("dump agent failed: {e:?}")
+            }
+        };
+
+        // We're going to collect just this region into an ersatz DumpAgentCore
+        (&mut udp_dump as &mut dyn DumpAgent).read_dump(
+            Some(DumpArea::ByIndex(dump_index as usize)),
+            &mut agent_core,
+            false,
+        )?;
+
+        // Pop the most recent dump, since we were just using it to read memory
+        // and it doesn't need to take up a dump area forever.
+        udp_dump.reinitialize_dump_from(dump_index)?;
```

Comment:
I don't know to what degree we care, but won't this also have the effect of popping the dump of any task that dumps in the interim?

Reply:
It will; I think that's a reasonable tradeoff in order to avoid filling up dump memory, but I'll add a comment to that effect.